.github/copilot-instructions.md

```diff
@@ -0,0 +1,238 @@
+- `ClusterRestProxyResource`: Proxies requests to Kafka/SR clusters
+
+OpenAPI specs auto-generated to `src/generated/resources/openapi.{yaml,json}`.
+
+### Connection State Management
+
+`ConnectionStateManager` is the single source of truth for connection lifecycle:
+- Uses CDI events for lifecycle notifications (Created, Updated, Connected, Disconnected, Deleted)
+- Each `ConnectionState` wraps a `ConnectionSpec` with runtime state
+- State changes fire events with qualifiers like `@Lifecycle.Connected` and `@ServiceKind.CCloud`
+
+**Observer pattern**: Other components inject `Event<ConnectionState>` to react to connection changes.
+
+## Development Workflows
+
+### Building & Testing
+
+**CRITICAL**: This project requires **Java 21** and **GraalVM CE**. Install via SDKMAN:
+```bash
+sdk env install  # Uses .sdkmanrc in project root
+```
+
+**Primary Make Commands:**
+- `make quarkus-dev`: Dev mode with hot reload (port 26636)
+- `make quarkus-test`: Continuous testing mode
+- `make build`: Full build (compiles, tests, packages)
+- `make test`: Run unit + integration tests (JVM mode)
+- `make test-native`: Build native executable and run integration tests against it
+- `make mvn-package-native`: Create native executable
+- `make mvn-generate-sidecar-openapi-spec`: Regenerate OpenAPI specs
+
+**Dev Mode URLs:**
+- Application: `http://localhost:26636`
+- Dev UI: `http://localhost:26636/q/dev-ui`
+- GraphQL UI: `http://localhost:26636/q/graphql-ui`
+- Swagger UI: `http://localhost:26636/swagger-ui`
+- OpenAPI spec: `http://localhost:26636/openapi`
+
+### Authentication in Development
+
+The sidecar uses token-based auth. Get a token via handshake:
+```bash
+export DTX_ACCESS_TOKEN=$(curl -s http://localhost:26636/gateway/v1/handshake | jq -r .auth_secret)
+```
+
+Use in requests: `Authorization: Bearer ${DTX_ACCESS_TOKEN}`
+
+In tests, use `@TestProfile(NoAccessFilterProfile.class)` to disable auth checks.
+
+### Testing Conventions
+
+**Unit tests** (`*Test.java`):
+- Use `@QuarkusTest` for CDI/injection support
+- Mock external dependencies
+- Focus on isolated component behavior
+
+**Integration tests** (`*IT.java`):
+- Annotated with `@Tag("io.confluent.common.utils.IntegrationTest")`
+- Use `@ConnectWireMock` for mocking external HTTP services (CCloud APIs, Kafka REST Proxy)
+- Run against real Quarkus-managed containers via DevServices
+- WireMock runs on `${quarkus.wiremock.devservices.port}` in tests
+
+**Native integration tests**:
+- Use profile `nativeit` (config in `src/test/resources/application-nativeit.yaml`)
+- Run with `make test-native`
+
+**Test utilities**: `SidecarClient` in `src/test/java/.../util/` provides helpers for GraphQL queries, connection creation, etc.
+
+## Project-Specific Patterns
+
+### Configuration
+
+All config in `src/main/resources/application.yml`, with profile overrides:
+- `%dev`: Development mode (port 26636, JSON logging off)
```

Comment:
wondering about adding something about conflicting ports since that is a common issue